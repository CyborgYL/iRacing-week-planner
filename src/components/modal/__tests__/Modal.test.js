import { describe, test } from '@jest/globals';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Modal from '../Modal';

describe('components/modal/Modal', () => {
  test('renders closed', async () => {
    const component = renderer.create(
      <Modal isOpen={false} onClose={() => {}} title="Test Closed">
        <div>Hello Closed!</div>
      </Modal>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders no footer', async () => {
    const component = renderer.create(
      <Modal isOpen onClose={() => {}} title="Test No Footer" showFooter={false}>
        <div>No Footer!</div>
      </Modal>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders open', async () => {
    const onClose = jest.fn(() => {});
    const doneAction = jest.fn(() => {});
    let component;

    const clickEvent = { preventDefault: () => {}, stopPropagation: () => {} };

    act(() => {
      component = renderer.create(
        <Modal
          isOpen
          onClose={onClose}
          title="Test Open"
          doneButtonText="Is Done"
          doneAction={doneAction}
          id="testOpen"
        >
          <div>Hello Open!</div>
        </Modal>,
      );
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root
        .findByProps({ 'aria-label': 'Close' })
        .props.onClick(clickEvent);
    });

    expect(onClose).toHaveBeenCalled();

    await act(async () => {
      await component.root
        .findByProps({ children: 'Is Done' })
        .props.onClick(clickEvent);
    });

    expect(doneAction).toHaveBeenCalled();
  });

  test('presses escape', async () => {
    const onClose = jest.fn(() => {});
    let component;
    act(() => {
      component = renderer.create(
        <Modal isOpen onClose={onClose} title="Test Escape" />,
      );
    });

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'KeyR', key: 82, keyCode: 82 }));
    expect(onClose).not.toHaveBeenCalled();

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'Escape', key: 27, keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(1);

    // Backwards compatibility
    document.dispatchEvent(new KeyboardEvent('keypress', { keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(2);

    act(() => {
      component.unmount();
    });

    document.dispatchEvent(new KeyboardEvent('keypress', { code: 'Escape', key: 27, keyCode: 27 }));
    expect(onClose).toHaveBeenCalledTimes(2);
  });

  test('default actions don\'t crash', async () => {
    let component;

    const clickEvent = { preventDefault: () => {}, stopPropagation: () => {} };

    act(() => { component = renderer.create(<Modal isOpen title="Test Open" doneButtonText="OK" />); });

    await act(async () => {
      await component.root.findByProps({ 'aria-label': 'Close' }).props.onClick(clickEvent);
    });

    await act(async () => {
      await component.root.findByProps({ children: 'OK' }).props.onClick(clickEvent);
    });
  });
});
