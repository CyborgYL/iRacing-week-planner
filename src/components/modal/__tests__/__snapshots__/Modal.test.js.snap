// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/Modal renders closed 1`] = `null`;

exports[`components/modal/Modal renders no footer 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Test No Footer
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div>
          No Footer!
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/Modal renders open 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    id="testOpen"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Test Open
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div>
          Hello Open!
        </div>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Is Done
        </button>
      </div>
    </div>
  </div>
</div>
`;
